File: scenarios/first-connect-bundle.ts
#!/usr/bin/env bun
// scenarios/first-connect-bundle.ts
//
// Headline demo for the productionized install flow: a fresh host
// (with helmor-server but no bundle) gets the agent runtime installed
// automatically on connect, and the Remote Servers row's live chip
// narrates what's happening end-to-end.
//
// Pre-record: the test container's `$HOME/.helmor/server/` has only
// `helmor-server` (the daemon binary). No sidecar, no claude, no
// wrapper script — the "I just SSH'd a stock Linux box" state.
//
// Captured beats (one scene per visual state):
//   1. The Remote Servers panel, freshly opened, showing the
//      connected runtime with its state chip but no agent-runtime
//      chip — the "before" baseline.
//   2. The chip turning blue with a spinner + the live phase
//      message ("Uploading agent runtime (3 files, 325.9 MB)") —
//      proof that the operator can see what's happening while it
//      happens, no opaque progress bar.
//   3. The chip green: "Agent runtime installed in 5.3s" — the
//      "done" beat that confirms a successful install.

import { Tape } from "./lib.ts";

const NAME = process.env.RUNTIME_NAME ?? "docker-linux-arm64";
const HOST = process.env.HOST_ALIAS ?? "helmor-taper-arm64";
const CONTAINER = process.env.CONTAINER ?? "helmor-test-linux-arm64";
const OUT = process.env.TAPE_DIR ?? "./tapes/first-connect-bundle";

const tape = new Tape("first-connect-bundle", OUT);
await tape.connect();

// 0. Wipe the bundle so this records a true cold install. Restoring
//    the daemon binary's name (the `.real → helmor-server` move) is
//    what makes the install treat this as fresh-host.
const wipe = Bun.spawn([
	"docker", "exec", "-u", "e2e", CONTAINER, "sh", "-c",
	"rm -f $HOME/.helmor/server/helmor-sidecar; " +
		"rm -f $HOME/.helmor/server/claude; " +
		"rm -f $HOME/.helmor/server/MANIFEST.json; " +
		"rm -rf $HOME/.helmor/server/.staging; " +
		"if [ -f $HOME/.helmor/server/helmor-server.real ]; then " +
		"  mv -f $HOME/.helmor/server/helmor-server.real $HOME/.helmor/server/helmor-server; " +
		"fi",
]);
if ((await wipe.exited) !== 0) {
	throw new Error("could not wipe container bundle artifacts");
}
tape.log("wiped container bundle artifacts");

// 1. Reload to a clean shell, then open Settings → Remote Servers.
//    Skip an explicit disconnect — `connect_remote_runtime` is
//    idempotent on the registry side and the install routine is
//    safe to re-trigger. We just need to land in the panel.
await tape.js('window.location.reload(); return "r";');
await tape.sleep(6000);
await tape.openSettings("remote-servers");
tape.assert("panel_opens", await tape.waitFor("[role=dialog]", 10_000));
tape.assert(
	"row_present",
	await tape.waitFor(`[data-testid=remote-server-row-${NAME}]`, 10_000),
);

await tape.scene({
	caption: `Remote Servers panel — ${NAME} is connected but has no agent runtime yet`,
	hold: 4,
});

// 2. Fire connect_remote_runtime in the background. We don't wait
//    for it to settle here — the whole point of this tape is to
//    capture the chip's progress, which only renders WHILE the
//    install is in flight. The chip's `installing` state lasts a
//    few seconds (the LAN install is ~5s), so the next scene
//    captures it mid-stream.
void tape.bridge.invokeCommand(
	"connect_remote_runtime",
	{ name: NAME, host: HOST, remoteBinary: "/home/e2e/.helmor/server/helmor-server", forwardAgent: false },
	"connect",
);

// Wait for the chip to flip into installing state.
const installingChip = await tape.waitFor(
	`[data-testid=remote-server-bundle-installing-${NAME}]`,
	10_000,
);
tape.assert("bundle_chip_installing", installingChip);

// Live-capture 3 s of the chip transitioning between phases (the
// recorder catches the actual frames; scene() runs ScreenCaptureKit
// in parallel with the UI animating).
await tape.scene({
	caption: `Auto-install: agent runtime streams to the container, sha-verified, atomic per file`,
	record: 4,
	hold: 6,
});

// 3. Wait for the install to settle into the success chip.
const installedChip = await tape.waitFor(
	`[data-testid=remote-server-bundle-installed-${NAME}]`,
	60_000,
);
tape.assert("bundle_chip_installed", installedChip);
const chipText = await tape.js<string | null>(
	`var c = document.querySelector('[data-testid=remote-server-bundle-installed-${NAME}]');` +
		`return c ? c.innerText : null;`,
);
tape.log(`chip says: ${chipText}`);

await tape.scene({
	caption: `Done — ${chipText ?? "agent runtime installed"}. Connect, install, ready in under 10 s.`,
	hold: 5,
});

const passed = await tape.finish({
	runtimeName: NAME,
	chipText,
});
process.exit(passed ? 0 : 1);

File: src/scenarios/first_connect_bundle.rs
//! Headline demo for the productionized install flow: a fresh host
//! (with helmor-server but no bundle) gets the agent runtime
//! installed automatically on connect, and the Remote Servers row's
//! live chip narrates what's happening end-to-end.
//!
//! Rust port of `scenarios/first-connect-bundle.ts`. The container
//! bundle is wiped via `docker exec` before recording so the install
//! beats are real cold-install transitions, not no-ops.

use std::process::Command;
use std::time::Duration;

use anyhow::{anyhow, Result};
use serde_json::{json, Value};

use crate::tape::{SceneSpec, Tape};

#[derive(Debug, Clone)]
pub struct Config {
    pub runtime_name: String,
    pub host_alias: String,
    pub container: String,
    pub remote_binary: String,
}

impl Config {
    pub fn from_env() -> Self {
        Self {
            runtime_name: std::env::var("RUNTIME_NAME")
                .unwrap_or_else(|_| "docker-linux-arm64".into()),
            host_alias: std::env::var("HOST_ALIAS")
                .unwrap_or_else(|_| "helmor-taper-arm64".into()),
            container: std::env::var("CONTAINER")
                .unwrap_or_else(|_| "helmor-test-linux-arm64".into()),
            remote_binary: std::env::var("REMOTE_BINARY")
                .unwrap_or_else(|_| "/home/e2e/.helmor/server/helmor-server".into()),
        }
    }
}

fn wipe_container_bundle(container: &str) -> Result<()> {
    // Mirror of the TS port's `docker exec ... sh -c "rm -f ...; mv -f ..."`.
    let script = "rm -f $HOME/.helmor/server/helmor-sidecar; \
                  rm -f $HOME/.helmor/server/claude; \
                  rm -f $HOME/.helmor/server/MANIFEST.json; \
                  rm -rf $HOME/.helmor/server/.staging; \
                  if [ -f $HOME/.helmor/server/helmor-server.real ]; then \
                    mv -f $HOME/.helmor/server/helmor-server.real $HOME/.helmor/server/helmor-server; \
                  fi";
    let out = Command::new("docker")
        .args(["exec", "-u", "e2e", container, "sh", "-c", script])
        .output()?;
    if !out.status.success() {
        return Err(anyhow!(
            "wipe bundle: docker exec exit {}: {}",
            out.status,
            String::from_utf8_lossy(&out.stderr).trim()
        ));
    }
    Ok(())
}

pub async fn run(tape: &mut Tape, config: &Config) -> Result<bool> {
    wipe_container_bundle(&config.container)?;
    tape.log("wiped container bundle artifacts");

    tape.js::<Value>(r#"window.location.reload(); return "r";"#).await?;
    tape.sleep(Duration::from_secs(6)).await;
    tape.open_settings("remote-servers").await?;
    let panel_opens = tape
        .wait_for("[role=dialog]", Duration::from_secs(10))
        .await?;
    tape.assert("panel_opens", panel_opens, "");
    let row_selector = format!("[data-testid=remote-server-row-{}]", config.runtime_name);
    let row_present = tape
        .wait_for(&row_selector, Duration::from_secs(10))
        .await?;
    tape.assert("row_present", row_present, "");
    tape.scene(
        SceneSpec::new(format!(
            "Remote Servers panel — {} is connected but has no agent runtime yet",
            config.runtime_name
        ))
        .hold_sec(4),
    )
    .await?;

    // Fire connect in the background so the chip can be captured mid-stream.
    tape.invoke_async(
        "connect_remote_runtime",
        json!({
            "name": config.runtime_name,
            "host": config.host_alias,
            "remoteBinary": config.remote_binary,
            "forwardAgent": false,
        }),
        "connect",
    )
    .await?;

    let installing_selector = format!(
        "[data-testid=remote-server-bundle-installing-{}]",
        config.runtime_name
    );
    let installing_chip = tape
        .wait_for(&installing_selector, Duration::from_secs(10))
        .await?;
    tape.assert("bundle_chip_installing", installing_chip, "");

    tape.scene(
        SceneSpec::new("Auto-install: agent runtime streams to the container, sha-verified, atomic per file")
            .record_sec(4)
            .hold_sec(6),
    )
    .await?;

    let installed_selector = format!(
        "[data-testid=remote-server-bundle-installed-{}]",
        config.runtime_name
    );
    let installed_chip = tape
        .wait_for(&installed_selector, Duration::from_secs(60))
        .await?;
    tape.assert("bundle_chip_installed", installed_chip, "");
    let chip_script = format!(
        r#"var c=document.querySelector({sel}); return c?c.innerText:null;"#,
        sel = serde_json::to_string(&installed_selector)?,
    );
    let chip_text: Option<String> = tape.js(&chip_script).await?;
    tape.log(&format!(
        "chip says: {}",
        chip_text.clone().unwrap_or_default()
    ));

    tape.scene(
        SceneSpec::new(format!(
            "Done — {}. Connect, install, ready in under 10 s.",
            chip_text
                .clone()
                .unwrap_or_else(|| "agent runtime installed".into())
        ))
        .hold_sec(5),
    )
    .await?;

    tape.finish(json!({
        "runtimeName": config.runtime_name,
        "chipText": chip_text,
    }))
    .await
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn config_from_env_defaults() {
        unsafe {
            std::env::remove_var("RUNTIME_NAME");
            std::env::remove_var("HOST_ALIAS");
            std::env::remove_var("CONTAINER");
            std::env::remove_var("REMOTE_BINARY");
        }
        let c = Config::from_env();
        assert_eq!(c.runtime_name, "docker-linux-arm64");
        assert_eq!(c.host_alias, "helmor-taper-arm64");
        assert_eq!(c.container, "helmor-test-linux-arm64");
        assert_eq!(c.remote_binary, "/home/e2e/.helmor/server/helmor-server");
    }
}
